### PR TITLE
feat: flow for creating a new workspace

### LINKED ISSUE #2437: Feature: Workspaces creation flow
Implement the workspaces onboarding flow. For reference, see @isabensusan's [work for onboarding in Figma](https://www.figma.com/file/sqRXS7hw6E93QgRMCl3WYz/Workspaces?type=design&node-id=1-57&mode=design&t=6MhgvGeqDbBP1wg4-0).

- [x] New workspaces layout
- [x] #2463
- [x] #2473
- [x] #2477
- [x] #2482
- [x] #2505

### LINKED ISSUE #2477: Create initial workspace creation flow (name, description and be able to delete)
The workspace creation flow is feature flagged sop we can roll out pieces of the creation flow starting off with creating a workspace with just a name and description, including the ability to delete a workspace.

### PR DESCRIPTION
**Pull request by nickytonline**
## Description

- Introduces a new Workspaces layout
- Page for creating a new workflow
- Introduces new routes:
	- `/worskpaces/new`
	- `/worskpaces/[workspaceId]/settings`

As a first phase since this is feature flagged, I'd like to get this merged as is and the tracking of repos will come in another PR @isabensusan, @brandonroberts. This way our team can start creating workspaces even if tracking repos or adding members isn't available yet.

It also avoids having a large PR since this piece is funcitonal on its own.

## What type of PR is this? (check all applicable)

- [x] 🍕 Feature
- [ ] 🐛 Bug Fix
- [ ] 📝 Documentation Update
- [ ] 🎨 Style
- [ ] 🧑‍💻 Code Refactor
- [ ] 🔥 Performance Improvements
- [ ] ✅ Test
- [ ] 🤖 Build
- [ ] 🔁 CI
- [ ] 📦 Chore (Release)
- [ ] ⏩ Revert

## Related Tickets & Documents
Closes #2477
Relates to #2437

## Mobile & Desktop Screenshots/Recordings

[CleanShot 2024-01-12 at 17 55 22@2x
](https://www.loom.com/share/58960cd1fb894a139df1f950d7416051?sid=be255e62-15bf-4738-950e-5fdfb0a026d7)

### Creating a workspace

1. Ensure you're logged in.
2. Go to https://deploy-preview-2461--oss-insights.netlify.app/workspaces/new
3. Enter a workspace name. If you don't and try to create the workspace, there should be client-side validation failing.
4. Add a description if you want. It's optional.
5. Click the create workspace button.
6. A toast message says the workspace was created and you are redirected to your workspace settings.
7. The workspace is deleted and you are redirected to the create new workspaces page.

### Editing a workspace

1. Select a workspace from the drop down in the sidebar
2. You're redirected to your workspace settings page, e.g. https://deploy-preview-2461--oss-insights.netlify.app/workspaces/some-workspace-id/settings
3. Change the name or description and save it.
4. A toast message says it was updated successfully.
5. At the bottom of the settings page, click on the Delete workspaces button.
6. A modal to confirm appears like our other destructive flows in the app.
9. Confirm deletion.
10. The workspace is deleted

Note: being redirected to the create new workspaces page after deletion is temporary. We'll redirect to most likely your landing page for your personal workspace in the future.

## Added to documentation?

- [ ] 📜 README.md
- [ ] 📓 docs.opensauced.pizza
- [ ] 🍕 dev.to/opensauced
- [ ] 📕 storybook
- [ ] 🙅 no documentation needed

### CODE REVIEW

File: middleware.ts [hunk 1]
```diff
@@ -13,7 +13,8 @@ const pathsToMatch = [
   "/feed/",
   "/user/notifications",
   "/user/settings",
-  "/account-deleted"
+  "/account-deleted",
+  "/workspaces/:path*",
```
Comment (nickytonline) [inline middleware.ts:17]: I'm curious how demo insights and demo lists play out in the context of workspaces @bdougie @isabensusan

**Review by jpmcb — COMMENTED** [review]
Overall, this is looking really great!! 👏🏼 

---

Maybe a missing hydration when updating the workspace name ... ?

If I create my workspace `woof woof` and then update the name to `Wooooof`, going to delete it doesn't seem to have persisted those changes:

Screenshot 2024-01-16 at 9 06 27 AM

Screenshot 2024-01-16 at 9 06 11 AM

**Comment by nickytonline** [comment]
> Overall, this is looking really great!! 👏🏼
> 
> Maybe a missing hydration when updating the workspace name ... ?
> 
> If I create my workspace `woof woof` and then update the name to `Wooooof`, going to delete it doesn't seem to have persisted those changes:
> 
> Screenshot 2024-01-16 at 9 06 27 AM Screenshot 2024-01-16 at 9 06 11 AM

I'll see what's up. Thanks John!

**Comment by bdougie** [comment]
@isabensusan mentioned the issue with me not seeing the feature flag was triaged and narrowed down to my user id in posthog being weird. I unfortunately QA this while its being the feature flag, and will lean others to review this one for now.

**Comment by bdougie** [comment]
> @isabensusan mentioned the issue with me not seeing the feature flag was triaged and narrowed down to my user id in posthog being weird. I unfortunately QA this while its being the feature flag, and will lean others to review this one for now.

correction. I can run locally with with the flag remove which I am doing now.

**Comment by nickytonline** [comment]
> Overall, this is looking really great!! 👏🏼
> 
> Maybe a missing hydration when updating the workspace name ... ?
> 
> If I create my workspace `woof woof` and then update the name to `Wooooof`, going to delete it doesn't seem to have persisted those changes:
> 
> Screenshot 2024-01-16 at 9 06 27 AM Screenshot 2024-01-16 at 9 06 11 AM

@bdougie @jpmcb, this is fixed in f62b396f60ec27aa84f3c6646c750fa0adaa5ffe.

**Comment by isabensusan** [comment]
looking great!!

File: components/shared/AppSidebar/AppSidebar.tsx [hunk 2]
```diff
@@ -49,8 +57,34 @@ export const AppSideBar = () => {
     .sort((a, b) => new Date(b.updated_at).getTime() - new Date(a.updated_at).getTime());
 
   return (
-    <div className="pt-12 bg-white flex flex-col gap-8 justify-between min-w-max min-h-full">
+    <div className="fixed top-0 pt-12 bg-white flex flex-col gap-8 justify-between min-w-max min-h-full">
       <div className="grid gap-4 mt-4 px-4">
+        <label className="flex flex-col gap-2 w-max">
+          <span className="sr-only">Workspace</span>
+          <ClientOnly>
+            <SingleSelect
```
Comment (nickytonline) [inline components/shared/AppSidebar/AppSidebar.tsx:65]: The values here don't update at the moment when someone edits the workspace settings. I'm looking to get this sorted today.

Comment (nickytonline) [reply components/shared/AppSidebar/AppSidebar.tsx:71]: Fixed in e48438b839370268bd7cf0f4b4fd4e39fd20d2a0

File: middleware.ts (hunk 1 above)
Comment (bdougie) [reply middleware.ts:18]: I'd lean towards showing demo or default workspace, i.e. topJS projects or am organization like vuejs as the default.

**Comment by bdougie** [comment]
update: I was able to test this locally by removing the feature flag in the middleware.ts

I did notice that I see other workspaces I did not create and get a 404 when viewing them. Most likely because I am not an admin. 

Screen Shot 2024-01-17 at 9 28 57 AM

**Comment by nickytonline** [comment]
> update: I was able to test this locally by removing the feature flag in the middleware.ts
> 
> 
> 
> I did notice that I see other workspaces I did not create and get a 404 when viewing them. Most likely because I am not an admin and the workspaces list is not limiting the spaces you can see. I have dug into the code though, just assuming.
> 
> 
> 
> Screen Shot 2024-01-17 at 9 28 57 AM
> 
> 

Those are workspaces I created. I'm not sure why you're seeing the. 🤔 The API sets the person who created the workspace as owner, but no one else atm. I'll chat with @jpmcb about this.

Maybe subduing in the API is missing or I'm not holding it right. 😅

**Comment by nickytonline** [comment]
I see that @jpmcb fixed a query so you should only see your own workspaces now @bdougie @brandonroberts.

**Comment by nickytonline** [comment]
This is good to go now. If there's tweaks @isabensusan, I can do them in a follow up PR or now. I leave it up to you to decide.

File: middleware.ts (hunk 1 above)
Comment (nickytonline) [reply middleware.ts:18]: Sounds good. It won't be part of this PR, but will probably come closer to the end of this whole feature.

**Review by bdougie — APPROVED** [review]
Flow works for me, also the confirming that I no longer see your workspaces.

**Review by brandonroberts — APPROVED** [review]

**Pull request merged**